doc/devguide/extending/app-layer/transactions.rst

```diff
@@ -0,0 +1,310 @@
+
+A TLS Handshake is a more complex example, where several messages are exchanged before the transaction is considered completed:
+
+.. image:: img/TlsHandshake.png
+  :width: 600
+  :alt: A sequence diagram with two entities, Client and Server, with an arrow going from the Client to the Server labeled "ClientHello" and below that an arrow going from Server to Client labeled "ServerHello". Below those arrows, several more follow from Server to Client and vice-versa, before a dotted line indicates that the transaction is finally completed.
+
+Template Protocol
+~~~~~~~~~~~~~~~~~
+
+Suricata has a template protocol for educational purposes, which has simple bidirectional transactions. 
+
+A completed transaction for the template looks like this:
+
+.. image:: img/TemplateTransaction.png
+  :width: 600
+  :alt: A sequence diagram with two entities, Client and Server, with an arrow going from the Client to the Server, labeled "Request". An arrow below that first one goes from Server to Client. 
+
+Following are the functions that check whether a transaction is considered completed, for the Template Protocol. Those are called by the Suricata API. Similar functions exist for each protocol, and may present implementation differences, based on what is considered a transaction for that given protocol.
+
+In C:
+
+.. code-block:: c
+
+    static int TemplateGetStateProgress(void *txv, uint8_t direction)
+    {
+        TemplateTransaction *tx = txv;
+    
+        SCLogNotice("Transaction progress requested for tx ID %"PRIu64
+            ", direction=0x%02x", tx->tx_id, direction);
+    
+        if (direction & STREAM_TOCLIENT && tx->response_done) {
+            return 1;
+        }
+        else if (direction & STREAM_TOSERVER) {
+            /* For the template, just the existence of the transaction means the
+             * request is done. */
+            return 1;
+        }
+    
+        return 0;
+    }
+
+And in Rust:
+
+.. code-block:: rust
+
+    pub extern "C" fn rs_template_tx_get_alstate_progress(
+        tx: *mut std::os::raw::c_void,
+        _direction: u8,
+    ) -> std::os::raw::c_int {
+        let tx = cast_pointer!(tx, TemplateTransaction);
+    
+        // Transaction is done if we have a response.
+        if tx.response.is_some() {
+            return 1;
+        }
+        return 0;
+    }
+
+Work In Progress changes
+========================
+
+Currently we are working to have files be part of the transaction instead of the per-flow state, as seen in https://redmine.openinfosecfoundation.org/issues/4444.
+
+Another work in progress is to limit the number of transactions per flow, to prevent Denial of Service by quadratic complexity (see  https://redmine.openinfosecfoundation.org/issues/4530).
```

Comment:
You can precise that this DOS is only for protocols which can have multiple transactions at the same time such as HTTP2 so-called streams...